### ISSUE TITLE
restic ls supports incremental listings of additional directories

### ISSUE THREAD

**Issue by garethgeorge** [restic/restic#4897]
What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

`restic ls` should support some mode (e.g. a `--paths-from-stdin`) flag where it allows continuous querying of paths within the repo without multiple CLI invocations. 

What I'm imagining here might look like support for reading new line delimited paths from stdin e.g. an example invocation might look like 

```
restic ls <snapshot ID> --json --paths-from-stdin
/
/foo/
/foo/bar/
/foo/bar/baz
```

with output
```
/foo
/foo/bar
/foo/otherdir
/foo/bar/baz
/foo/bar/moredirsinbaz
````

e.g. the user is able to incrementally provide additional query paths, for each filter restic rewalks the tree and outputs any relevant data. 

What are you trying to do? What problem would this solve?
---------------------------------------------------------

My aim here is to introduce a CLI capability that minimizes followup listing latency when performing sequential queries e.g. to support snapshot browsers built on top of the CLI. 

A snapshot browser implementation might look something like. `restic ls <snapshot ID> /` to list files in the top directory, and then to repeatedly invoke `restic ls <snapshot ID> /foo` `restic ls <snapshot ID> /foo/bar/` `restic ls <snapshot ID> /foo/bar/baz`. 

It would be nice to be able to reduce the latency of those followup listings by keeping the restic process alive between listings and by instead sending additional queries to the same CLI process.

Alternatives might be simply listing the whole repo and implementing filtering logic on top of this but the latency here can be unacceptable for VERY large repos with huge sets of small files. `restic ls <snapshot ID> <dirs>` today seems to be much faster than listing the whole repo.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

restic has been a phenomenal tool and provides the most and performant foundation for backups I've found anywhere (across an incredible number of platforms). Really impressed with the project for a long time and I've been having a lot of fun building https://github.com/garethgeorge/backrest on top of it, would love to contribute back features that add capabilities useful for backrest and wrappers like it.

Design Ideas / Proposal
-------------------------------------------------------------

My initial reading makes me think a simple loop around https://github.com/restic/restic/blob/master/cmd/restic/cmd_ls.go#L362-L404 would be an easy first pass on this idea and would allow skipping most of the initialization work which I suspect makes followup calls to ls slow today.  

Wanting to get some general input on this approach / alternatives I may not be seeing and I'd be interested to followup this FR with some benchmarks / possibly a PR.

**Comment by garethgeorge** [comment]
Spent some time prototyping this after work, posting some numbers to illustrate the gains:
```
repository 5ff2f593 opened (version 2, compression level auto)
[0:00] 100.00%  2 / 2 index files loaded
snapshot 58f1f94b of [/home/dontpanic/Documents/github/other/restic] at 2024-07-03 17:21:50.039543588 -0700 PDT by dontpanic@pop-os filtered by [/doc]:
... dropped output ...
./restic ls latest /doc  0.47s user 0.02s system 70% cpu 0.687 total
```

**Comment by MichaelEischer** [comment]
I don't want to slow your enthusiasm, but I'm wondering whether some proper API would be better suited for your needs. Something like https://github.com/restic/restic/pull/4754 maybe?

Introducing wildly different options to add interactivity for special cases to different commands sounds rather brittle to me. In terms of CLI design it feels a bit weird to add an option to the `ls` that is specifically designed for usage with wrappers. It's also not particularly extensible. What if you want to switch to a different snapshot? Or notice during the operation that more metadata for each file would be necessary?

**Comment by garethgeorge** [comment]
re: WebDAV. I think it seems reasonable is probably the ideal solution in terms of the capabilities that can be built upon it, that said restic separately already supports parsable JSON output over the CLI for wrappers and that approach has been very effective for my purposes for orchestration. It's a perfectly decent API for restic and the only place it really falls over is flows that want to be interactive (which is really just listings that I've found so far).

tl;dr I'd like to follow along with WebDAV and if there's any effort that I can lend to accelerate that along that may be a good option, but I'm also interested to think through low cost options that might still be inline with restic's usage philosophy AND that allows for quick iterative exploration on the CLI. 

Many unix commands support piped input so I'm curious if there's a useful way we can define that for `ls` but understand if that's not inline with restic's design -- I don't think any other commands really accept pipe'd input at the moment.

**Comment by MichaelEischer** [comment]
> Many unix commands support piped input so I'm curious if there's a useful way we can define that for `ls` but understand if that's not inline with restic's design -- I don't think any other commands really accept pipe'd input at the moment.

Besides reading the repository password from stdin, the backup command is the only one that (optionally) reads from stdin. The big conceptual difference is that so far stdin/stdout is "data" only, there are no commands passed along this path. Properly handling such commands will require some sort of RPC protocol. (It is for example perfectly valid for filenames to contain linebreaks (!) on Unix systems, which would break splitting stdin by linebreaks).

Making use of WebDAV would sort of provide that for free as we want to provide that interface anyways.

**Comment by garethgeorge** [comment]
Hey, following up on this thread after a few months as I've seen a bit more discussion online re: file browsing in Backrest (and restic in general across projects e.g. restic-browser and other alternatives) as a pain point (e.g. [here on reddit](https://www.reddit.com/r/unRAID/comments/1b4z743/comment/lj3sml2/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button), https://github.com/garethgeorge/backrest/issues/228).

It's interesting to think through what would make sense for the input format. I think keeping the stdin "data like" would be the right way to support this e.g. semantically a stream of new line delimited [strconv.Quote](https://pkg.go.dev/strconv#Quote) escaped strings as a "path list" or such.

It makes sense to worry re: maintainability that this isn't an RPC mechanism so it's not very extensible, but I also can't imagine much more one would want re: options that would need to change with low latency (i.e. where it isn't acceptible to kill and fork another restic process and use flags to reconfigure). The major blocker for a fast file browser via cli automation (e.g. competitive with the browsers in projects like kopia, duplicati, etc) is just a supported `ls` loop supporting a way to incrementally drill into directories.

The last consideration I think is that, extending `ls` is not as full featured as WebDAV _but_ a stdin / stdout interface can be automated in the same style as existing restic commands which use a JSON over stdout format for automation. From my perspective, there's something to be said for the ease of wrapping the stdin / stdout solution given that it's already a problem that anyone scripting restic needs to solve for other commands. WebDAV requires me to manage a WebDAV server's lifecycle and poll until I know it's available / listening.

**Comment by MichaelEischer** [comment]
> I think keeping the stdin format very simple e.g. simply a datafile that's compatible with essentially (cat <my paths to list.txt)

The absolutely simplest format would be separating the input paths with null bytes `\0`.  File paths are not allowed to contain null bytes, so that would allow omitting any encoding. Similarly the text output lines would be separated by null bytes, and three consecutive null bytes indicate the end of the snapshot. To signal an error the output would be two consecutive null bytes, then the error message and a final null byte "\0\0{some JSON error message}\0".

That format would work for both text and JSON output. Although for scripting purposes I'm somewhat tempted to always use JSON output. I assume you'd expect to use `ls` without the recursive parameter?

> containing a list of [strconv.Quote](https://pkg.go.dev/strconv#Quote) escaped strings as a "path list" or such.

Once we start quoting things, we should just use JSON to not introduce yet another encoding format. The most minimal data format for the request would be a plain JSON string, although I still prefer adding a command like the following `{"command":"ls", "path":"/example", "snapshot":"12345678"}`. Then the output format would naturally also be JSON.
Probably plus some additional way to signal errors.
(I've thought about suggesting JSON-RPC, but that doesn't support streaming the result data)

When looking at your prototype implementation, I've noticed that the end of the snapshot seems to be indicated using `Listing took %v`. That would also have to be replaced by an end of list message, or maybe just an empty JSON message.

> It makes sense to worry re: maintainability that this isn't an RPC mechanism so it's not very extensible, but I also can't imagine much more one would want re: options that would need to change with low latency (i.e. where it isn't acceptible to kill and fork another restic process and use flags to reconfigure).

The reason I keep bringing up this more extensible variant is that it's basically guaranteed that there will be feature requests to also have such an interactive behavior for further commands. As those are likely to work on the same repository, it would make sense to open the repository only once and then interact with it. What I'd particularly like to avoid is littering the codebase with various extensions for interactivity that have to be maintained in addition to a then combined command.

Does backrest work with arbitrary restic versions or do you bundle a specific version? I'd be willing to give an interactive ls implementation a try while gating it behind a feature flag. Once it becomes clear how an interactive scripting interface will look like, we can either stabilize it or deprecate and eventually remove it.

**Comment by garethgeorge** [comment]
> Once we start quoting things, we should just use JSON to not introduce yet another encoding format. The most minimal data format for the request would be a plain JSON string, although I still prefer adding a command like the following {"command":"ls", "path":"/example", "snapshot":"12345678"}.  ...

That definitely sounds workable from my perspective. And right -- I don't think recursive flag makes much sense here (and in fact the caller could implement it themself with the new capability...). WDYT re: flag naming in that case to make it abundantly clear that this is a scriptability feature? 

> When looking at your prototype implementation, I've noticed that the end of the snapshot seems to be indicated using Listing took %v. That would also have to be replaced by an end of list message, or maybe just an empty JSON message.

This can be pulled out for the PR, that is just instrumentation to verify the speedup.

> Does backrest work with arbitrary restic versions or do you bundle a specific version? I'd be willing to give an interactive ls implementation a try while gating it behind a feature flag. Once it becomes clear how an interactive scripting interface will look like, we can either stabilize it or deprecate and eventually remove it.

Backrest downloads the public `restic` from releases and checks signature today, but I don't think bundling a specific build (especially if bringing backrest specific features) is something that the community would balk at. A feature flag is definitely a workable approach while we figure this out.


Overall if the direction sounds workable happy to put out a PR for more detailed discussion, I've also left a comment on https://github.com/restic/restic/pull/4754 and am happy to lend effort over there if that makes more sense / is close.

**Comment by MichaelEischer** [comment]
> WDYT re: flag naming in that case to make it abundantly clear that this is a scriptability feature?

Hmm, I'd somewhat like to avoid "interactive" as part of the flag name as users would expect that they can use `ls` interactively. My current favorite is `--script-mode` which both signals that it significantly changes the command's behavior and that it is intended for scripting usage.

> Backrest downloads the public `restic` from releases and checks signature today, but I don't think bundling a specific build (especially if bringing early features) is something that the community would balk at.

So you're essentially just using the latest release? With feature flags, the interactive mode would also be part of the officially released restic version, although hidden behind the feature flag. The feature flag approach will however mean that a certain backrest version probably only supports restic versions in a certain range. restic versions that are far newer than backrest would probably not work (although that risk always exists).

> Overall if the direction sounds workable happy to put out a PR for more detailed discussion

The code for a scripting mode should be reasonably separate from the remaining code of the `ls` command. That way we don't have to untangle it in the future. Reusable code should be shared nevertheless.

> I've also left a comment on #4754 and am happy to lend effort over there if that makes more sense / is close.

Good question, I don't really know how far along the cleanups for that PR are. What is sure is that the code can't life in `cmd/restic` (there's already far too much code there that belongs into some other package), so this will require at least a major refactoring.

I also haven't made my mind up yet regarding how an interactive variant of restic should look like. While simple extensions for each command can get quite far, it would be even more powerful to have a full-blown RPC/REST interface. The problem is that those add even more code to maintain. What I definitely want to avoid is having a wild mix of interfaces.
